src/compiler/parser.ts

```diff
@@ -1001,10 +1001,10 @@ module ts {
             switch (parsingContext) {
                 case ParsingContext.SourceElements:
                 case ParsingContext.ModuleElements:
-                    return isSourceElement(inErrorRecovery);
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isModuleElement();

```

Comment:
This is very cryptic and difficult to read.  Can you please extract this into a helper function that makes the flow more clear.  For example, i would expect it to look something like:

```ts
function isSourceOrModuleElement(inErrorRecovery: boolean) {
    // Explanation of this case so people understand why we do this.
    if (token === SyntaxKind.Semicolon && inErrorRecovery) {
        return false;
    }

    return isModuleElement();
}
```